### ISSUE TITLE
Add support to 2D Matryoshka 

### ISSUE THREAD

**Issue by juanluisdb** [UKPLab/sentence-transformers#2504]
There is a [new paper](https://arxiv.org/abs/2402.14776) that adapt Matryoshka to also work with the number of layers required. 
I think it could be interesting to add it to the supported losses. I could work on it.

The technique is simple, at every training step, they compute the loss in the same way that Matryoshka, then, sample a random layer to also compute the loss using that output layer.

For training, `output_hidden_states`  should be set to True and using a code similar to the [Matryoshka](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/MatryoshkaLoss.py) loss should work. 

Other details should be considered:

- For loading the model, a new parameter should be sent to set the number of layers and model is loaded with the specified layers.

Any thoughts on this?

**Comment by tomaarsen** [comment]
Hello!

Actually, if I remember right from reading the paper, there is also a KL divergence loss to try and make the embeddings at the sampled layer match that of the final embeddings. But yes, I think this paper is fascinating, and I would love to see it implemented in ST as soon as possible.

Now, my thinking was actually to separate the implementation from the current MatryoshkaLoss. With other words, rather than just implementing `Matryoshka2DLoss` or something, I was hoping to implement only the `AdaptiveLayerLoss` or something like that, which accepts another loss:

```python
class AdaptiveLayerLoss(nn.Module):
    def __init__(self, model, loss):
         ...
```

Then, the `AdaptiveLayerLoss` gets a provided loss function `base_loss`:
1. samples a layer $n$ (or gets all layers? after all, why only sample one?)
2. computes the `base_loss` over the final embeddings
3. computes the `base_loss` over embeddings at layer $n$
4. computes KL-div loss between the layer $n$ embeddings and the final embeddings.
5. The losses from 2, 3 and 4 are added together to produce the final loss.

Then, we can create a shorthand:
```python
class Matryoshka2DLoss(AdaptiveLayerLoss):
    def __init__(
        self,
        model: SentenceTransformer,
        loss: nn.Module,
        matryoshka_dims: List[int],
        matryoshka_weights: Optional[List[Union[float, int]]] = None,
    ) -> None:
        super().__init__(model=model, loss=MatryoshkaLoss(model=model, loss=loss, matryoshka_dims=matryoshka_dims, matryoshka_weights=matryoshka_weights)
```

And this should be roughly equivalent to the 2d Matryoshka paper.

That way, we implement both 2d Matryoshka, but also the functionality that you can easily only take the first $n$ layers that can be used completely separately from the `MatryoshkaLoss`, e.g. just like: 
```python
base_loss = CoSENTLoss(model)
loss = AdaptiveLayerLoss(model, loss=base_loss)
```
Because I'm sure some people are interested in speeding up their inference, but don't care to get dynamically smaller embeddings (i.e. they don't care about "normal" Matryoshka).

I'm very curious to hear your thoughts on this @juanluisdb 

I suspect that I will start working on this tomorrow, and hopefully this can be released in 1 or 2 weeks. It will also take some changes to the `SentenceTransformer` class to allow only partially using a model.

Also, cc the first author of the 2D Matryoshka paper, in case they're interested in following this work @SeanLee97

- Tom Aarsen

**Comment by juanluisdb** [comment]
That solution would be great!

Regarding the number of layers to use, I haven't found explicit mention in the paper of the reason, but I suspect it could be to make the training easier. However, I agree that ST should be flexible and allow the users to decide how they want to proceed. To do that, `AdaptiveLayerLoss`  could accept some parameters:

```
class AdaptiveLayerLoss(nn.Module):
    def __init__(self, model, loss, do_sample: bool, n_layers_per_step: int):
         ...
```

Best

**Comment by juanluisdb** [comment]
yeah, the original MRL computes the losses for all the sizes: https://github.com/RAIVNLab/MRL/blob/main/MRL.py#L23

It would be nice to hear from @SeanLee97 👀

**Comment by ir2718** [comment]
This discussion is somewhat related to what I was thinking about suggesting, especially the part about using a model up to layer $n$. I've recently been doing some research on exploiting prior transformer layers for sentence similarity and I found this paper really cool: [BERT Has More to Offer: BERT Layers Combination Yields Better Sentence Embeddings](https://aclanthology.org/2023.findings-emnlp.1030.pdf). To summarize, what they do is they take combinations of layer outputs and apply mean pooling, then use a validation set to find out which combination performs the best. 

I would like to propose a few things to make adding 2D Matryoshka more optimized for inference and more flexible for implementing other methods that use different layers:
1. The parameter sent would be a list of indices of layers
    - If a model has 12 layers, and you want to use all of them you would use `list(range(13))` (usually input embeddings are treated as a layer)
    - The default setting would be using `[-1]`
2. If the max value in the list of indices is smaller than the number of layers, then using `output_hidden_states=True` is unncessary
    - This isn't important for fine-tuning in 2D Matryoshka as it uses all layers (if I'm not mistaken), but would be a great thing for inference
    - ideally, you would only do the forward pass up to the maximum index value in the provided list of indices
    - I've also found that fine-tuning a model using a sentence embedding from a prior layer can be an easy and flexible way to obtain a smaller model 
3. If the max value in the list of indices is smaller than the number of layers, then you don't need to load all the layers
    - This would be a nice thing for inference, as it would take up less memory

**Comment by SeanLee97** [comment]
> Hello!
> 
> Actually, if I remember right from reading the paper, there is also a KL divergence loss to try and make the embeddings at the sampled layer match that of the final embeddings. But yes, I think this paper is fascinating, and I would love to see it implemented in ST as soon as possible.
> 
> Now, my thinking was actually to separate the implementation from the current MatryoshkaLoss. With other words, rather than just implementing `Matryoshka2DLoss` or something, I was hoping to implement only the `AdaptiveLayerLoss` or something like that, which accepts another loss:
> 
> ```python
> class AdaptiveLayerLoss(nn.Module):
>     def __init__(self, model, loss):
>          ...
> ```
> 
> Then, the `AdaptiveLayerLoss` gets a provided loss function `base_loss`:
> 
> 1. samples a layer n (or gets all layers? after all, why only sample one?)
> 2. computes the `base_loss` over the final embeddings
> 3. computes the `base_loss` over embeddings at layer n
> 4. computes KL-div loss between the layer n embeddings and the final embeddings.
> 5. The losses from 2, 3 and 4 are added together to produce the final loss.
> 
> Then, we can create a shorthand:
> 
> ```python
> class Matryoshka2DLoss(AdaptiveLayerLoss):
>     def __init__(
>         self,
>         model: SentenceTransformer,
>         loss: nn.Module,
>         matryoshka_dims: List[int],
>         matryoshka_weights: Optional[List[Union[float, int]]] = None,
>     ) -> None:
>         super().__init__(model=model, loss=MatryoshkaLoss(model=model, loss=loss, matryoshka_dims=matryoshka_dims, matryoshka_weights=matryoshka_weights)
> ```
> 
> And this should be roughly equivalent to the 2d Matryoshka paper.
> 
> That way, we implement both 2d Matryoshka, but also the functionality that you can easily only take the first n layers that can be used completely separately from the `MatryoshkaLoss`, e.g. just like:
> 
> ```python
> base_loss = CoSENTLoss(model)
> loss = AdaptiveLayerLoss(model, loss=base_loss)
> ```
> 
> Because I'm sure some people are interested in speeding up their inference, but don't care to get dynamically smaller embeddings (i.e. they don't care about "normal" Matryoshka).
> 
> I'm very curious to hear your thoughts on this @juanluisdb
> 
> I suspect that I will start working on this tomorrow, and hopefully this can be released in 1 or 2 weeks. It will also take some changes to the `SentenceTransformer` class to allow only partially using a model.
> 
> Also, cc the first author of the 2D Matryoshka paper, in case they're interested in following this work @SeanLee97
> 
> * Tom Aarsen

Thanks for considering integrating 2DMSE 🪆🪆 into SentenceTransformers! I would like to help.

**Comment by tomaarsen** [comment]
@SeanLee97 
Could you share the hyperparameters used in the various losses?
![image](https://github.com/UKPLab/sentence-transformers/assets/37621491/5c4b2231-a27b-4ee6-ae9e-ba8b2adf7508)

I can't find that information in the paper.

- Tom Aarsen

**Comment by SeanLee97** [comment]
@tomaarsen 

1) We currently set 1.0 for all \lambda.
2) Did you experiment with AnglE (https://github.com/SeanLee97/AnglE) or ST? If use ST, could you share the code?

**Comment by tomaarsen** [comment]
>     1. We currently set 1.0 for all \lambda.

Thank you, that clears it up!

>     2. Did you experiment with AnglE (https://github.com/SeanLee97/AnglE) or ST? If use ST, could you share the code?

Custom code for ST. I'll make a branch so you can have a look at it. I'm currently testing it out with MNRL (i.e. in batch negatives) and the "Adaptive Layer" that I described above, rather than "full" 2DMSE which also uses the adaptable output dimensionality.

Here are my findings so far:
![image](https://github.com/UKPLab/sentence-transformers/assets/37621491/5d1f7d54-6860-42f8-b539-347ad650c9d4)

1. `distilroberta-base-nli-adaptive-layer-no-kl`: Sum of MNRL loss for each layer
2. `distilroberta-base-nli-adaptive-layer-kl-weighted`: MNRL loss for the final embedding + (sum of MNRL loss for each non-final layer / `num_layers` - 1) + (sum of KL-div for each non-final layer / (`num_layers` - 1)).
    This is kind of like "normal" loss at weight 1, loss at non-final layers at weight 1 and KL-div loss at non-final layers at weight 1.
3. `distilroberta-base-nli-adaptive-layer-kl-sampled`: MNRL loss for the final embedding + MNRL loss at a randomly sampled layer + KL-div at a randomly sampled layer.
4. `distilroberta-base-nli-adaptive-layer-no-kl-weighted`: MNRL loss for the final embedding + (sum of MNRL loss for each non-final layer / `num_layers` - 1)
5. `distilroberta-base-nli`: Normal MNRL, no Adaptive Layer loss, no Matryoshka loss

- Tom Aarsen
